kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/-MoshiKotlinExtensions.kt

```diff
@@ -0,0 +1,89 @@
+package com.squareup.moshi.kotlin.reflect
+
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import com.squareup.moshi.internal.NonNullJsonAdapter
+import com.squareup.moshi.internal.NullSafeJsonAdapter
+import java.lang.reflect.Type
+import kotlin.reflect.KClass
+import kotlin.reflect.KType
+import kotlin.reflect.KTypeParameter
+import kotlin.reflect.KTypeProjection
+import kotlin.reflect.KVariance
+import kotlin.reflect.typeOf
+
+/**
+ * @return a [JsonAdapter] for [T], creating it if necessary. Note that while nullability of [T]
+ *         itself is handled, nested types (such as in generics) are not resolved.
+ */
+@ExperimentalStdlibApi
+inline fun <reified T> Moshi.adapter(): JsonAdapter<T> {
+  return adapter(typeOf<T>())
+}
+
+@ExperimentalStdlibApi
+inline fun <reified T> Moshi.Builder.addAdapter(adapter: JsonAdapter<T>) = add(typeOf<T>().toType(), adapter)
```

Comment:
@rharter @swankjesse Thought experiment - if `T` is nullable, should `.nullSafe()` be called on the adapter? and/or vice-versa with `.nonNull()`?

Reply:
I'm curious about the use case there.  I don't use the nullSafe variants much since I lean on the properties to handle that.  Would the user have to do this if the class can be used in different contexts with different nullability?

```
Moshi.Builder()
  .addAdapter(JsonAdapter<Foo?>())
  .addAdapter(JsonAdapter<Foo>())
```

Reply:
hmm, yeah. And moshi wouldn't be able to tell the difference between those two because they're both just assigned to `Foo`, so I think we have to just leave it at that for now